Convert nano::network periodic functions to use a steady timer instead of callbacks

clemahieu opened the pull request:
This change pulls in the boost coroutine library into the node.

Additionally, it rewrites the ongoing_cleanup function in nano::network in terms of steady timers, which is easier to reason around instead of callback loops through nano::alarm.

clemahieu committed:
This change pulls in the boost coroutine library in to the node.
Additionally it rewrites the ongoing_cleanup function in nano::network in terms of steady timers, which is easier to reason around instead of callback loops through nano::alarm.

guilhermelawless commented on nano/node/network.cpp:
```diff
@@ -727,14 +731,18 @@ void nano::network::cleanup (std::chrono::steady_clock::time_point const & cutof
+	node.io_ctx,
+	[this](boost::asio::yield_context yield) {
+		boost::system::error_code ec;
+		while (!stopped && !ec)
```
Should we debug_assert at the end that `!ec || ec == boost::system::errc::operation_canceled` ?

guilhermelawless commented on nano/node/network.cpp:
```diff
@@ -727,14 +731,18 @@ void nano::network::cleanup (std::chrono::steady_clock::time_point const & cutof
 		}
-	});
+	},
+	boost::coroutines::attributes (128 * 1024));
```
Is this a leftover?

guilhermelawless commented on nano/node/network.cpp:
```diff
@@ -132,6 +135,7 @@ void nano::network::stop ()
 		{
 			thread.join ();
 		}
+		cleanup_timer.cancel ();
```
Does this operation to be queued in a strand?
Also think we should use the overload with error_code to swallow exceptions.

clemahieu replied:
The documentation says it kills any asynchronous operation and the error_code variant is deprecated.
https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/basic_waitable_timer/cancel.html

clemahieu committed:
This change pulls in the boost coroutine library in to the node.
Additionally it rewrites the ongoing_cleanup function in nano::network in terms of steady timers, which is easier to reason around instead of callback loops through nano::alarm.

guilhermelawless reviewed (approved):
This introduces a thread sanitizer segfault on daemon shutdown.

On inspection it does not look like an issue with the implementation. It happens when multiple threads run the io context and stackful coros were spawned. It appears the segfault happens after all I/O is handled which makes it a non-issue.

wezrule reviewed (approved):
LGTM
